strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/repositories/GremlinRepository.java

```diff
@@ -0,0 +1,139 @@
+
+import org.apache.tinkerpop.gremlin.structure.Element;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.carlspring.strongbox.data.domain.DomainObject;
+import org.carlspring.strongbox.gremlin.adapters.EntityTraversalAdapter;
+import org.carlspring.strongbox.gremlin.adapters.UnfoldEntityTraversal;
+import org.carlspring.strongbox.gremlin.dsl.EntityTraversal;
+import org.carlspring.strongbox.gremlin.dsl.EntityTraversalSource;
+import org.carlspring.strongbox.gremlin.tx.TransactionContext;
+import org.neo4j.ogm.session.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.repository.CrudRepository;
+
+/**
+ * @author sbespalov
+ *
+ */
+@Transactional
+public abstract class GremlinRepository<S extends Element, E extends DomainObject> implements CrudRepository<E, String>
+{
+
+    private static final Logger logger = LoggerFactory.getLogger(GremlinRepository.class);
+
+    @Inject
+    @TransactionContext
+    private Graph graph;
+    @Inject
+    protected Session session;
+    
+    protected EntityTraversalSource g()
+    {
+        return graph.traversal(EntityTraversalSource.class);
+    }
+
+    protected abstract EntityTraversal<S, S> start(E entity, Supplier<EntityTraversalSource> g);
+    
+    protected abstract EntityTraversal<S, S> start(Supplier<EntityTraversalSource> g);
+
+    public <R extends E> R save(Supplier<EntityTraversalSource> g, R entity) {
+        String uuid = merge(g, entity);
+        
+        return (R) findById(g, uuid).get();    
+    }
+    
+    public abstract String merge(Supplier<EntityTraversalSource> g, E entity);
+    
+    public Optional<E> findById(Supplier<EntityTraversalSource> g, String uuid){
+        String label = adapter().label();
+        EntityTraversal<S, E> traversal = start(g).findById(uuid, label)
+                                                  .map(adapter().fold());
+        if (!traversal.hasNext())
+        {
+            return Optional.empty();
+        }
+
+        return Optional.of(traversal.next());        
+    }
+    
+    public Optional<E> findById(String uuid)
+    {
+        return findById(this::g, uuid);
+    }
+
+    @Override
+    public <S extends E> Iterable<S> saveAll(Iterable<S> entities)
+    {
+        throw new UnsupportedOperationException("TODO implement");
+    }
+
+    @Override
+    public boolean existsById(String id)
+    {
+        throw new UnsupportedOperationException("TODO implement");
+    }
+
+    @Override
+    public Iterable<E> findAll()
+    {
+        throw new UnsupportedOperationException("TODO implement");
+    }
+
+    @Override
+    public Iterable<E> findAllById(Iterable<String> ids)
+    {
+        throw new UnsupportedOperationException("TODO implement");
+    }
+
+    @Override
+    public long count()
+    {
+        throw new UnsupportedOperationException("TODO implement");
+    }
```

Comment:
Do we need follow-up tasks for these?
If "yes", could you please create tasks for these? Thanks! :)

Reply:
will do

Reply:
Thanks! :)

Reply:
[issues/1844](https://github.com/strongbox/strongbox/issues/1844)

Reply:
Thanks for creating the follow-up task.